doc/sphinx/user-extensions/syntax-extensions.rst

```diff
@@ -7,64 +7,177 @@ In this chapter, we introduce advanced commands to modify the way Coq
 parses and prints objects, i.e. the translations between the concrete
 and internal representations of terms and commands.
 
-The main commands to provide custom symbolic notations for terms are
-:cmd:`Notation` and :cmd:`Infix`; they will be described in the
-:ref:`next section <Notations>`. There is also a
-variant of :cmd:`Notation` which does not modify the parser; this provides a
-form of :ref:`abbreviation <Abbreviations>`. It is
-sometimes expected that the same symbolic notation has different meanings in
-different contexts; to achieve this form of overloading, Coq offers a notion
-of :ref:`notation scopes <Scopes>`.
-The main command to provide custom notations for tactics is :cmd:`Tactic Notation`.
+When inputing a term to Coq, it goes through several successive steps.
+
+* lexing: the input is split in a linear stream of tokens, for
+  instance the string ``"1 + 2 * 2"`` will be split into the sequence
+  of five tokens `'1' '+' '2' '*' '2'`
+
+* parsing: the previous stream is given a tree-like structure, for
+  instance here
+
+  .. code-block:: text
+     :name: after-parsing
+
+             *
+            / \
+           1   +
+              / \
+             2   2
+
+* notation interpretation: each syntactic element gets translated to a
+  term, for instance :n:`1` can be interpreted as the natural number
+  :n:`S O` then :n:`2` is interpreted as :n:`S (S O)`, then :n:`2 * 2`
+  as :n:`Nat.mul (S (S O)) (S (S O))` and finally our whole term as
+  :n:`Nat.add (S O) (Nat.mul (S (S O)) (S (S O)))`.
```

Comment:
This only applies to translating single token notations (and maybe only numbers).

Multitoken notations are implemented by adding additional rules to the parser grammar.  They are not implemented as operations on the parse tree.  For example,

```
(* in Notations.v )
Reserved Notation "x <= y" (at level 70, no associativity).
Reserved Notation "x < y" (at level 70, no associativity).
Reserved Notation "x >= y" (at level 70, no associativity).
Reserved Notation "x > y" (at level 70, no associativity).
```

generates the following productions in the grammar:

```
Print Grammar.

(*
Entry term is
  :
| "70" RIGHTA
  [ SELF; ">"; NEXT
  | SELF; ">="; NEXT
  | SELF; "<"; NEXT
  | SELF; "<="; NEXT
*)
```

There is additional post-parsing logic that distinguishes based on the notation scope (e.g. '+' in nat vs Z vs Reals)

Reply:
> This only applies to translating single token notations (and maybe only numbers).

Could you elaborate? In what sense are "x + y" or "x * y" single token notations? Maybe the example would be clearer with variables rather than constants? or mixing both?

> Multitoken notations are implemented by adding additional rules to the parser grammar. They are not implemented as operations on the parse tree. For example,
[...]
> generates the following productions in the grammar:
[...]

Sure but I'm not sure the detailled syntax of camlp5 is of any interest in this introductory paragraph. That syntax only appears through the `Print Grammar` which is a pretty advanced command (and even more so now that we have `Print Notation`). That syntax is documented there but I'm not sure it is worth introducing it anywhere else.

> There is additional post-parsing logic that distinguishes based on the notation scope (e.g. '+' in nat vs Z vs Reals)

Indeed, that paragrah doesn't even mention scopes. Maybe it should?